### PR TITLE
Build nightly `pydantic` version for `Pydantic >= 2.0` test build

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Removes the hacky monkeypatch workaround and instead fully tests Pydantic with the pickling fixes included.

Upon the next Pydantic release, we can remove the logic to build from source and just install `2.5.0`.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ci/docker/serve.build.Dockerfile [hunk 1]
```diff
@@ -26,5 +26,8 @@ RUN pip install -U --ignore-installed \
 
 RUN git clone https://github.com/wg/wrk.git /tmp/wrk && pushd /tmp/wrk && make -j && sudo cp wrk /usr/local/bin && popd
 
-# Install custom Pydantic version if requested.
-RUN if [[ -z $PYDANTIC_VERSION ]] ; then echo Not installing custom Pydantic version ; else pip install -U pydantic==$PYDANTIC_VERSION ; fi
+# Install Pydantic from source if requested.
+RUN if [[ -z $PYDANTIC_BUILD_COMMIT ]] ; \
+    then echo "Not installing Pydantic from source." ; \
+    else git clone https://github.com/pydantic/pydantic && pushd pydantic && git checkout $PYDANTIC_BUILD_COMMIT && pip install -e . && popd ; \
```
Comment (GeneDer) [inline ci/docker/serve.build.Dockerfile:32]: Non-blocker: We can probably just do ` pip install -U git+https://github.com/pydantic/pydantic@$PYDANTIC_BUILD_COMMIT` to install on a particular commit 🙃

**Review by GeneDer — APPROVED** [review]
:shipit:

File: ci/docker/serve.build.Dockerfile (hunk 1 above)
Comment (edoakes) [reply ci/docker/serve.build.Dockerfile:32]: Oh, you're totally right...

**Review by aslonnie — APPROVED** [review]

File: ci/docker/servepydantic2.build.wanda.yaml [hunk 2]
```diff
@@ -6,6 +6,6 @@ srcs:
   - python/requirements_compiled.txt
   - python/requirements/test-requirements.txt
 build_args:
-  - PYDANTIC_VERSION=2.4.2
+  - PYDANTIC_BUILD_COMMIT=60c5db6e1ea55d4e5fc13234810d513b3b1b03ae
```
[the remark this answers is not in the document]
Comment (edoakes) [reply ci/docker/servepydantic2.build.wanda.yaml:9]: yea will do

**Pull request merged**